### PR TITLE
fix(module-loader) - Fixed ModuleLoader behavior with multiple modulecatalogs

### PR DESCRIPTION
**Pull request by lilla28**
Changes:
- picked up changes from branch and commits related to the ModuleLoader when it tries to register its ModuleCatalog https://github.com/morganstanley/ComposeUI/commit/89ec2d9e1f9be47df0cacfdbb4586d9405d0b332 && https://github.com/morganstanley/ComposeUI/commit/57ad7daab2f8e28d63b335562b7f8b2bdf20fe79
- fixed tests when it registers its modulecatalogs (mocking array of catalogs)

Please refer to the PR https://github.com/morganstanley/ComposeUI/pull/569 that I accidentally closed.
I copied all the review comments/suggestions that the PR received and resolved them if they were marked as Resolved.
Also please be aware that the previous PR is not merged into any branch!

### CODE REVIEW

File: src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/AggregateModuleCatalog.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+// or implied. See the License for the specific language governing permissions
+// and limitations under the License.
+
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace MorganStanley.ComposeUI.ModuleLoader;
```
Comment (lilla28) [inline src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/AggregateModuleCatalog.cs:19]: [ZKRobi](https://github.com/ZKRobi) [2 weeks ago](https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1542820209)
Please consider moving this class to it's own file as it implements complex behavior. I'd also consider unit testing it, although integration testing may cover all use-cases. That said, I don't see the multi-catalog scenario covered by tests, and I don't see what happens if multiple catalogs provide an entry with the same moduleId (perhaps with different contents).

File: architecture/adr-014-module-catalog.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# Architecture Decision Record: Multiple ModuleCatalogs
+
+## Context
+ComposeUI aims to support many different kind of modules. These modules, that can be started by the ModuleLoader should be stored in ModuleCatalogs and every module should have its unique module id. The user should have the ability to handle multiple ModuleCatalogs with the ModuleLoader.
+
+General problem statement:
+- When multiple ModuleCatalogs are used, some modules could have the same module id in different ModuleCatalogs which could cause issues while starting or stopping the module via the ModuleLoader.
+
+## Decision
+- If the same module id is being used for multiple modules (likely in different ModuleCatalogs), the first occurrence will be used. In case of multiple ModuleCatalog registrations, the order of the registration determines which module can be used.
```
Comment (lilla28) [inline architecture/adr-014-module-catalog.md:17]: [ZKRobi](https://github.com/ZKRobi) [4 days ago](https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1555944683)
The same id in the same catalog shouldn't ever exist, although it's fine to describe the behavior.
I'd specify what is the first occurance in case of multiple catalogs though -> e.g. in the order of provider registration sounds like a method that is easy to understand, and provides some agency to the developer. This might have implications on how to implement merging modulecatalogs in case of async fetching.

File: architecture/adr-014-module-catalog.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<!-- Morgan Stanley makes this available to you under the Apache License, Version 2.0 (the "License"). You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0. See the NOTICE file distributed with this work for additional information regarding copyright ownership. Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License. -->
+
+| Type          | ID            | Status        | Title                   |
+| ------------- | ------------- | ------------- | ----------------------- |
+| ADR           | adr-014       | Accepted      | Multiple ModuleCatalogs |
```
Comment (lilla28) [inline architecture/adr-014-module-catalog.md:5]: [ZKRobi](https://github.com/ZKRobi) [4 days ago](https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1555945826)
I'd suggest discussing and putting the ADR to Accepted state before merging it together with the PR

[lilla28](https://github.com/lilla28) [3 days ago](https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1557297627)
I will create a new commit with the change of the status once everything is resolved and we discussed.

File: src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/AggregateModuleCatalog.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+        return _moduleIdToCatalog.Keys;
+    }
+
+    private readonly Lazy<Task> _lazyEnumerateModules;
+    private readonly ConcurrentDictionary<string, IModuleCatalog> _moduleIdToCatalog = new();
+
+    private Task EnumerateModules() => _lazyEnumerateModules.Value;
+
+    private async Task EnumerateModulesCore()
```
Comment (lilla28) [inline src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/AggregateModuleCatalog.cs:56]: [ZKRobi](https://github.com/ZKRobi) [4 days ago] --Old version is async fetching using LINQ.(https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1555987809)
This piece of code seems correct, but it's quite hard to understand. I'd separate it to fetch the moduleIds in parallel (it might even be more efficient that way), concatenate the results in order of registration, and do the resolving logic on this.

File: src/module-loader/dotnet/tests/MorganStanley.ComposeUI.ModuleLoader.Tests/AggregateModuleCatalogTests.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    [InlineData("noModuleId")]
+    [InlineData("testModuleId8")]
+    public async Task GetManifest_ThrowsModuleNotFoundException_WhenNoModuleIdFound(string moduleId)
+    {
+        var action = () => _moduleCatalog.GetManifest(moduleId);
+        await action.Should().ThrowAsync<ModuleNotFoundException>();
+    }
+
+    [Theory]
+    [InlineData("testModuleId")]
```
Comment (lilla28) [inline src/module-loader/dotnet/tests/MorganStanley.ComposeUI.ModuleLoader.Tests/AggregateModuleCatalogTests.cs:66]: [ZKRobi](https://github.com/ZKRobi) [4 days ago](https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1555999304)
The same Id exists with the same name multiple times. This doesn't verify the original source for the manifest as it can return any of the occurances and the assertion will stay true.

File: architecture/adr-014-module-catalog.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+# Architecture Decision Record: Multiple ModuleCatalogs
+
+## Context
+ComposeUI aims to support many different kind of modules. These modules, that can be started by the ModuleLoader should be stored in ModuleCatalogs and every module should have its unique module id. The user should have the ability to handle multiple ModuleCatalogs with the ModuleLoader.
+
+General problem statement:
+- When multiple ModuleCatalogs are used, some modules could have the same module id in different ModuleCatalogs which could cause issues while starting or stopping the module via the ModuleLoader.
+
+## Decision
```
Comment (lilla28) [inline architecture/adr-014-module-catalog.md:16]: [ztanczos](https://github.com/ztanczos) [3 days ago](https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1557497203)
Just an idea: what if we'd fail fast & hard if there are duplicates, i.e.: throw an exception and crash? What do you think?

[lilla28](https://github.com/lilla28) [3 days ago](https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1557597237)
Yeah we talked about it as well last week we talked about that it would be a kind of a weird thing to throw here an exception so we decided to go with the first item.

[ZKRobi](https://github.com/ZKRobi) [3 days ago](https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1557771342)
To recap, the issue with failing on duplicates from different sources is that it might cause a situation that the developer/user has no way of resolving.
Ideally this is never an issue but if merging multiple module catalogs is a feature we want to support and we wish to keep the standard interface (and not do something like e.g. prefixing ids that would be non-standard), we need a well documented consistent way of resolving these conflicts that the developer/user has a way to control.

File: src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/AggregateModuleCatalog.cs (hunk 1 above)
Comment (lilla28) [reply src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/AggregateModuleCatalog.cs:19]: Resolved as it was resolved on the original PR.

File: src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/AggregateModuleCatalog.cs (hunk 4 above)
Comment (lilla28) [reply src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/AggregateModuleCatalog.cs:56]: Resolved as it was resolved on the original PR.

File: src/module-loader/dotnet/tests/MorganStanley.ComposeUI.ModuleLoader.Tests/AggregateModuleCatalogTests.cs (hunk 5 above)
Comment (lilla28) [reply src/module-loader/dotnet/tests/MorganStanley.ComposeUI.ModuleLoader.Tests/AggregateModuleCatalogTests.cs:66]: Resolved as it was resolved on the original PR.

**Review by ztanczos — APPROVED** [review]

**Review by ZKRobi — APPROVED** [review]

**Review by kruplm — APPROVED** [review]
